src/main/java/hudson/plugins/ec2/EC2AbstractSlave.java

```diff
@@ -109,6 +113,11 @@
     /* The time at which we fetched the last instance data */
     protected transient long lastFetchTime;
 
+    /** Terminate was scheduled */
+    @GuardedBy("this")
+    @Nonnull
+    protected transient volatile CountDownLatch terminateScheduled = new CountDownLatch(0);
```

Comment:
A volatile lock? This seems overly complicated now. Do you think [CyclicBarrier](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/CyclicBarrier.html) would work better here? If you need to be able to reset the latch, that would work better with the barrier.

Reply:
CyclicBarrier makes this worse due to `BrokenBarrierExceptions` that can occur.

One thing i can do is move the nasty bits into a separate utility class and make a ResettableCountdownLatch

Reply:
That sounds like a better approach. Go for it!